### PR TITLE
Fix TypeError for ManagedIdentityCredential when using service principal credentials

### PR DESCRIPTION
**Pull request by M3GH4NN**
### What issues does this PR fix or reference?
Fixes: #46 

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [X] Docs
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: src/saltext/azurerm/utils/azurerm.py [hunk 1]
```diff
@@ -100,13 +100,20 @@ def _determine_auth(**kwargs):
...
     try:
         credentials = DefaultAzureCredential(authority=authority, **kwargs)
-    except ClientAuthenticationError:
-        raise SaltInvocationError(  # pylint: disable=raise-missing-from
-            "Unable to determine credentials. "
-            "A subscription_id with username and password, "
-            "or client_id, secret, and tenant or a profile with the "
-            "required parameters populated"
-        )
+    except:  # pylint: disable=W0702
```
Comment (nicholasmhughes) [inline src/saltext/azurerm/utils/azurerm.py:103]: **issue (blocking):** try to get away from the bare except here.

File: src/saltext/azurerm/utils/azurerm.py [hunk 2]
```diff
@@ -100,13 +100,20 @@ def _determine_auth(**kwargs):
...
+        if "client_id" and "tenant" and "secret" in kwargs:
+            credentials = get_identity_credentials(**kwargs)
+        else:
+            kwargs.pop("client_id")
```
Comment (nicholasmhughes) [inline src/saltext/azurerm/utils/azurerm.py:107]: **issue (blocking):** if we're modifying the input kwargs, we should log that we're doing so... even if it's in debug

**issue (blocking):** the preceding logic doesn't check if `client_id` exists in kwargs, so it's possible we throw a `KeyError` here.

File: src/saltext/azurerm/utils/azurerm.py [hunk 3]
```diff
@@ -332,19 +338,22 @@ def get_identity_credentials(**kwargs):
...
             os.environ[value] = kwargs[keyword]
-
     try:
         if kwargs.get("cloud_environment") and kwargs.get("cloud_environment").startswith("http"):
             authority = kwargs["cloud_environment"]
         else:
             authority = getattr(
                 KnownAuthorities, kwargs.get("cloud_environment", "AZURE_PUBLIC_CLOUD")
             )
-        log.debug("AUTHORITY: %s", authority)
```
Comment (nicholasmhughes) [inline src/saltext/azurerm/utils/azurerm.py:343]: **question (blocking):** why remove this log message?

File: src/saltext/azurerm/utils/azurerm.py [hunk 4]
```diff
@@ -100,13 +100,20 @@ def _determine_auth(**kwargs):
...
-            "or client_id, secret, and tenant or a profile with the "
-            "required parameters populated"
-        )
+    except:  # pylint: disable=W0702
+        if "client_id" and "tenant" and "secret" in kwargs:
+            credentials = get_identity_credentials(**kwargs)
+        else:
+            kwargs.pop("client_id")
+            try:
+                credentials = DefaultAzureCredential(authority=authority, **kwargs)
```
Comment (nicholasmhughes) [inline src/saltext/azurerm/utils/azurerm.py:109]: **suggestion (non-blocking):** I'm not a huge fan of this modified workflow, not from an operational perspective, but a code structure perspective. It seems a lot like "let's try to auth. it failed, so let's try another way. that also failed, so let's try another way". It _feels_ like you could do some checking up front in order to "do the right thing" on the first try.

**Review by nicholasmhughes — CHANGES_REQUESTED** [review]

File: src/saltext/azurerm/utils/azurerm.py (hunk 3 above)
Comment (M3GH4NN) [reply src/saltext/azurerm/utils/azurerm.py:343]: Oops, thought it was a debug statement of mine. Added back.

File: src/saltext/azurerm/utils/azurerm.py [hunk 5]
```diff
@@ -99,14 +99,18 @@ def _determine_auth(**kwargs):
...
+            credentials = DefaultAzureCredential(authority=authority, **kwargs)
     except ClientAuthenticationError:
-        raise SaltInvocationError(  # pylint: disable=raise-missing-from
+        raise SaltInvocationError(
             "Unable to determine credentials. "
             "A subscription_id with username and password, "
             "or client_id, secret, and tenant or a profile with the "
             "required parameters populated"
-        )
+        ) from None
```
Comment (nicholasmhughes) [inline src/saltext/azurerm/utils/azurerm.py:112]: **issue (blocking):** if the `raise` is just going to be from `None`, just put the pylint disable statement back and remove the `from None`

File: src/saltext/azurerm/utils/azurerm.py [hunk 6]
```diff
@@ -174,7 +178,6 @@ def get_client(client_type, **kwargs):
             base_url=cloud_env.endpoints.resource_manager,
             user_agent_policy=user_agent,
         )
-
```
Comment (nicholasmhughes) [inline src/saltext/azurerm/utils/azurerm.py:177]: **suggestion (non-blocking):** unless a `pre-commit` check like `black` is adding/removing empty lines, it's best to refrain from unrelated whitespace changes in PRs.

File: src/saltext/azurerm/utils/azurerm.py [hunk 7]
```diff
@@ -332,7 +335,6 @@ def get_identity_credentials(**kwargs):
     for keyword, value in kwarg_map.items():
         if kwargs.get(keyword):
             os.environ[value] = kwargs[keyword]
-
```
Comment (nicholasmhughes) [inline src/saltext/azurerm/utils/azurerm.py:335]: **suggestion (non-blocking):** unless a `pre-commit` check like `black` is adding/removing empty lines, it's best to refrain from unrelated whitespace changes in PRs.

File: src/saltext/azurerm/utils/azurerm.py [hunk 8]
```diff
@@ -341,10 +343,16 @@ def get_identity_credentials(**kwargs):
                 KnownAuthorities, kwargs.get("cloud_environment", "AZURE_PUBLIC_CLOUD")
             )
         log.debug("AUTHORITY: %s", authority)
+
```
Comment (nicholasmhughes) [inline src/saltext/azurerm/utils/azurerm.py:346]: **suggestion (non-blocking):** unless a `pre-commit` check like `black` is adding/removing empty lines, it's best to refrain from unrelated whitespace changes in PRs.

**Review by nicholasmhughes — CHANGES_REQUESTED** [review]

**Review by nicholasmhughes — CHANGES_REQUESTED** [review]
**chore (blocking):** looks good. just need a changelog file

File: changelog/46.fixed.md [hunk 9]
```diff
@@ -0,0 +1 @@
+Added try/except logic for Azure Authentication methods.
```
Comment (nicholasmhughes) [inline changelog/46.fixed.md:1]: **issue (blocking):** this is one of the things that you _did_... what issue did you _fix_?

**Example:** Fixed TypeError for ManagedIdentityCredential when using service principal credentials.

**Review by nicholasmhughes — CHANGES_REQUESTED** [review]

**Review by nicholasmhughes — APPROVED** [review]
